File: LeetCode/1011. Capacity To Ship Packages Within D Days.py
# https://leetcode.com/problems/capacity-to-ship-packages-within-d-days/description/

from test import test


class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> float:
        max_weight = max(weights)
        sum_of_weights = sum(weights)

        start = max_weight
        end = sum_of_weights

        min_capacity = float("inf")

        while start <= end:
            current_capacity = (end + start) // 2

            current_sum = 0

            days_count = 0

            for weight in weights:
                if current_sum + weight > current_capacity:
                    days_count += 1
                    current_sum = 0

                current_sum += weight

            if current_sum != 0:
                days_count += 1

            if days_count <= days:
                min_capacity = min(min_capacity, current_capacity)

                end = current_capacity - 1
                continue

            start = current_capacity + 1

        return min_capacity


solution = Solution()


test(
    solution.shipWithinDays,
    [
        {
            "input": [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5],
            "expected": 15,
        },
        {
            "input": [[3, 2, 2, 4, 1, 4], 3],
            "expected": 6,
        },
        {
            "input": [[1, 2, 3, 1, 1], 4],
            "expected": 3,
        },
        {
            "input": [[3, 2, 2, 4], 3],
            "expected": 4,
        },
    ],
)

File: LeetCode/1011. Capacity To Ship Packages Within D Days.js
// https://leetcode.com/problems/capacity-to-ship-packages-within-d-days/description/

const { test } = require('./test');

const shipWithinDays = (weights, days) => {
  // Минимальная вместимость (capacity) равна самому большому весу в массиве weights
  // Максимальная вместимость равна сумме всех весов (если надо перевезти все за один день)
  let sum = 0;

  let max = Number.NEGATIVE_INFINITY;

  for (let index = 0; index < weights.length; index++) {
    sum += weights[index];
    max = Math.max(max, weights[index]);
  }

  let start = max;
  let end = sum;

  let lastCorrect = Number.POSITIVE_INFINITY;

  // Бинарным поиском находим середину интервала
  while (start <= end) {
    const capacity = Math.floor((start + end) / 2);

    // Проверяем, за сколько дней можно с текущей capacity перевезти все грузы
    let currentSum = 0;
    let daysCount = 0;

    for (let index = 0; index < weights.length; index++) {
      currentSum += weights[index];

      if (currentSum < capacity) {
        continue;
      }

      daysCount++;

      if (currentSum === capacity) {
        currentSum = 0;
      } else {
        currentSum = weights[index];
      }
    }

    if (currentSum !== 0) {
      daysCount++;
    }

    // Если мы можем сделать это за целевое количество дней, то сохраняем capacity
    // И теперь можно проверить, что будет если capacity уменьшить
    if (daysCount <= days) {
      lastCorrect = Math.min(capacity, lastCorrect);

      end = capacity - 1;

      continue;
    }

    // Если не можем, то проверяем для большей capacity
    start = capacity + 1;
  }

  return lastCorrect;
};

test(shipWithinDays, [
  {
    input: [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5],
    expected: 15,
  },
  {
    input: [[3, 2, 2, 4, 1, 4], 3],
    expected: 6,
  },
  {
    input: [[1, 2, 3, 1, 1], 4],
    expected: 3,
  },
  {
    input: [[3, 2, 2, 4], 3],
    expected: 4,
  },
]);
